packages/backend-function/src/function_client_config_generator.ts

```diff
@@ -0,0 +1,128 @@
+  constructor(env: Env) {
+    this.env = env;
+  }
+
+  async getCredentialsAndIdentityId(): Promise<
+    CredentialsAndIdentityId | undefined
+  > {
+    const credentials: CredentialsAndIdentityId = {
+      credentials: {
+        accessKeyId: this.env.AWS_ACCESS_KEY_ID,
+        secretAccessKey: this.env.AWS_SECRET_ACCESS_KEY,
+        sessionToken: this.env.AWS_SESSION_TOKEN,
+      },
+    };
+    return credentials;
+  }
+  clearCredentialsAndIdentityId(): void {}
+}
+
+type Env = {
+  AWS_ACCESS_KEY_ID: string;
+  AWS_SECRET_ACCESS_KEY: string;
+  AWS_SESSION_TOKEN: string;
+  AMPLIFY_DATA_GRAPHQL_ENDPOINT?: string;
+  AWS_REGION: string;
+};
+
+/**
+ * Configures the amplify library for lambda execution with data-schema client configuration
+ *
+ * @param env The env vars provided for lambda execution
+ * @param modelIntrospection The MIS generated for the schema
+ * @returns the two named inputs to use with \`Amplify.configure(resourceConfig, libraryOptions)\`
+ */
+function lambdaConfig(env: Env, modelIntrospection: object) {
+  const customCredentialsProvider = new CustomCredentialsProvider(env);
+
+  return {
+    resourceConfig: {
+      API: {
+        GraphQL: {
+          endpoint: env.AMPLIFY_DATA_GRAPHQL_ENDPOINT ?? "",
+          region: env.AWS_REGION,
+          defaultAuthMode: "iam" as const,
+          modelIntrospection: modelIntrospection as any,
+        },
+      },
+    },
+    libraryOptions: {
+      Auth: {
+        credentialsProvider: customCredentialsProvider,
+      },
+    },
+  };
+};
+const { resourceConfig, libraryOptions } = lambdaConfig(
+  env,
+  modelIntrospection
+);
+
+export { resourceConfig, libraryOptions };
+`;
+
+/**
+ * Generates a client config for environment variables
+ */
+export class FunctionClientConfigGenerator {
+  private readonly header =
+    '// This file is auto-generated by Amplify. Edits will be overwritten.';
+
+  private clientConfigFilePath: string;
+
+  private indentation: string = '  ';
+
+  /**
+   * Initialize client config file name and location
+   */
+  constructor(private readonly functionName: string) {
+    this.clientConfigFilePath = `${process.cwd()}/.amplify/generated/client-config/${
+      this.functionName
+    }.ts`;
+  }
+  /**
+   * Generate a client configuration
+   */
+  generateClientConfig = () => {
+    this.writeFile(
+      [this.header, lambdaConfigHelper(this.functionName)].join(EOL)
+    );
+  };
+
+  /**
+   * Check if the script content has a client-config import
+   * @param scriptContent script content to check
+   * @returns true if the script content has a client-config import, false otherwise
+   */
+  usedBy(scriptContent: string) {
+    // Matches an import of client-config
+    const clientConfigImportRegex = /import.*from.*client-config/g;
+    return clientConfigImportRegex.test(scriptContent);
+  }
+
+  private writeFile = (content: string) => {
+    const clientConfigFileDirname = path.dirname(this.clientConfigFilePath);
```

Comment:
This pattern is copied from the env shim generation. Is it worth abstracting out the shared parts to keep the code cleaner and ensure that any updates to the write pattern only need be performed in one place?

Reply:
I'd follow the rule of three here and wait for more examples.

It's hard to say if this is true or incidental duplication.
https://anthonysciamanna.com/2018/07/28/the-dry-principle-and-incidental-duplication.html

Reply:
We have 3 if we keep the MIS json separate, but having removed that file just now, we only have 2 instances so I will skip this refactor for now.